File: Binary Search/Python/Split-Array-Largest-Sum.py
 // Brute force TC: O(n * maxLimit - minLimit )SC:O(1)
class Solution:
    def splitArray(self, a: List[int], k: int) -> int:
        if(k>len(a)): return -1
        minLim = max(a)
        maxLim = sum(a)
        for i in range(minLim,maxLim):
            if self.check(a,i,k): return i
        return -1
    
    def check(self,a,val,k):
        cnt=1
        allo = 0
        for i in range(len(a)):
            if allo + a[i] > val:
                allo = a[i]
                cnt=cnt+1
            else:
                allo += a[i]
        return cnt<=k

// optimal approach : TC:O(n⋅log(sum(a)−max(a))) SC:O(1)  
class Solution:
    def splitArray(self, a: List[int], k: int) -> int:
        if k > len(a):
            return -1
        low = max(a)
        high = sum(a)
        while low<=high:
            mid = low+(high-low) // 2
            if self.check(a,mid,k): high=mid-1
            else : low=mid+1
        return low
    
    def check(self,a,val,k):
        cnt=1
        allo = 0
        for i in range(len(a)):
            if allo + a[i] > val:
                allo = a[i]
                cnt=cnt+1
            else:
                allo += a[i]
        return cnt<=k

File: Binary Search/Split-Array-Largest-Sum.cpp
 // Brute force TC: O(n * maxLimit - minLimit )SC:O(1)
class Solution {
public:
    bool check(vector<int>& a, int val, int k) {
        int cnt = 1;
        int allocated = 0;
        for (int n : a) {
            if (allocated + n > val) {
                allocated = n;
                cnt++;
            } else
                allocated += n;
        }
        return (cnt <= k);
    }
    int splitArray(vector<int>& a, int k) {
        if (k > a.size())
            return -1;
        int midLim = *max_element(a.begin(), a.end());
        int maxlim = accumulate(a.begin(), a.end(), 0);
        for (int i = midLim; i <= maxlim; i++) {
            if (check(a, i, k))
                return i;
        }
        return -1;
    }
};

// optimal approach : TC:O(n⋅log(sum(a)−max(a))) SC:O(1)  
class Solution {
public:
    bool check(vector<int>& a, int val, int k) {
        int cnt = 1;
        int allocated = 0;
        for (int n : a) {
            if (allocated + n > val) {
                allocated = n;
                cnt++;
            } else
                allocated += n;
        }
        return (cnt <= k);
    }
    int splitArray(vector<int>& a, int k) {
        if (k > a.size())
            return -1;
        int low = *max_element(a.begin(), a.end());
        int high = accumulate(a.begin(), a.end(), 0);
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (check(a, mid, k))
                high = mid - 1;
            else
                low = mid + 1;
        }
        return low;
    }
};
